### PR TITLE
feat: Add basic authentication

### PR DESCRIPTION
**Pull request by Paula-Kli**
Refactors and rebuilds parts of #882

### CODE REVIEW

**Comment by Paula-Kli** [comment]
I think I found most accurances of the RoleVerification and could replace it by the `get_username` auth_injectable
what I am still wondering about:

- `/backend/capellacollab/core/authentication/provider/azure/routes.py` l.62, l.82
-  `/backend/capellacollab/core/authentication/provider/oauth/keystore.py` l.74
-  `/backend/capellacollab/core/authentication/provider/oauth/routes.py` l.31
- `/backend/capellacollab/core/logging/__init__.py` l.62

then is it fine to leave:
`/backend/capellacollab/sessions/routes.py` l.276 as I don't know whether it makes sense to replace it to also use BasicAuth
and what about
`/backend/tests/conftest.py` l.78?

**Comment by MoritzWeber0** [comment]
> I think I found most accurances of the RoleVerification and could replace it by the `get_username` auth_injectable what I am still wondering about:

> * `/backend/capellacollab/core/authentication/provider/azure/routes.py` l.62, l.82
> * `/backend/capellacollab/core/authentication/provider/oauth/keystore.py` l.74
> * `/backend/capellacollab/core/authentication/provider/oauth/routes.py` l.31

These three are OAuth related. They have a token by design, there is no need to add the abstraction. 

> * `/backend/capellacollab/core/logging/__init__.py` l.62

This has to be replace as we don't necessarily have a token here anymore.

> then is it fine to leave: `/backend/capellacollab/sessions/routes.py` l.276 as I don't know whether it makes sense to replace it to also use BasicAuth 

There should be no token anymore in general routes. We don't have a token when we're performing basic authentication, so the route would break. 

> and what about `/backend/tests/conftest.py` l.78?

`JWTBearer` is still called when we use OAuth as far as I can see. So we'd still need the fixture.

File: frontend/src/app/general/auth/basic_auth_service/basic_auth_token.service.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/*
```
Comment (MoritzWeber0) [inline frontend/src/app/general/auth/basic_auth_service/basic_auth_token.service.ts:1]: Please use the Angular CLI to create new components and services. Generated components never have a `_` in their filename, Angular uses `-` instead.

File: frontend/src/app/general/auth/basic_auth_token/basic_auth_token.component.html [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+<div style="margin-bottom: 10px; align-items: center">
+  <h2>Token Overview</h2>
+</div>
+<div class="input-container" style="align-items: center; margin-bottom: 10px">
+  <input
+    type="text"
+    [(ngModel)]="tokenDescription"
+    placeholder="Token Description"
+    style="margin-right: 10px"
+  />
```
Comment (MoritzWeber0) [inline frontend/src/app/general/auth/basic_auth_token/basic_auth_token.component.html:15]: Please use Angular Material input fields instead: https://material.angular.io/components/categories

**Review by MoritzWeber0 — COMMENTED** [review]
This is only a first review, mostly focusing on UI and UX of the feature. 
I haven't looked into the code deeply yet, so this has to be addressed in another review.

File: frontend/src/app/app.module.ts [hunk 3]
```diff
@@ -44,6 +44,7 @@ import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
 import { CookieModule } from 'ngx-cookie';
```
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply frontend/src/app/app.module.ts:44]: fixed this with @dominik003

File: backend/capellacollab/sessions/hooks/t4c.py [hunk 4]
```diff
@@ -81,14 +81,15 @@ def configuration_hook(
 
         for repository in t4c_repositories:
             try:
+                token_username = auth_injectables.get_username(token)
```
Comment (dominik003) [inline backend/capellacollab/sessions/hooks/t4c.py:84]: I would prefer simply naming this `username` here as we do it at all other places. 
In addition, we can't simply use it like this. The `get_username` function expects a fastapi request and not the token. So I would simply inject the username instead of the token in the `request_persistent_session` route and change the configuration hook to expect the username directly and not the token.

File: backend/capellacollab/users/tokens/crud.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        .all()
+    )
+
+
+def delete_token(
+    db: orm.Session, existing_token: models.DatabaseUserTokenModel
+) -> models.DatabaseUserTokenModel:
+    db.delete(existing_token)
+    db.commit()
+    return existing_token
```
Comment (dominik003) [inline backend/capellacollab/users/tokens/crud.py:66]:
```suggestion
```
We just deleted the token, so I would not return it here.

**Review by dominik003 — COMMENTED** [review]
I was mainly looking at the backend changes as I have already seen comments for the UI from Moritz. Looks really good, only some minor comments from my side.

File: frontend/src/app/general/auth/basic_auth_token/basic_auth_token.component.html [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  <div *ngIf="password" style="margin-bottom: 10px">
+    <div>Generated Password: {{ password }}</div>
+    <button
+      class="font-mono cursor-copy"
+      [cdkCopyToClipboard]="password"
+      (click)="showClipboardMessage()"
+    >
+      <mat-icon>content_copy</mat-icon>
+    </button>
+  </div>
```
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply frontend/src/app/general/auth/basic_auth_token/basic_auth_token.component.html:34]: image

The Gitlab UI also looks good. Especially the comment "Make sure you save it - you won't be able to access it again." is important.

File: backend/capellacollab/core/authentication/basic_auth.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        if not credentials:
+            if self.auto_error:
+                raise fastapi.HTTPException(
+                    status_code=status.HTTP_401_UNAUTHORIZED,
+                    detail="Not authenticated",
+                    headers={"WWW-Authenticate": "Basic"},
+                )
+            return None
+        with database.SessionLocal() as session:
+            user = user_crud.get_user_by_name(session, credentials.username)
```
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply backend/capellacollab/core/authentication/basic_auth.py:33]: we do through the "Token invalid for user" Exception if that is what you meant if the user is None l. 38

[the remark this answers is not in the document]
Comment (dominik003) [reply backend/capellacollab/core/authentication/basic_auth.py:33]: I think when I reviewed this the `if user` check to get the `token_data` was not there yet. Therefore, the `user.id` to get the `token_data` produced a valid mypy error on my side. But with l. 35 we indeed through an error in case the user is None

[the remark this answers is not in the document]
Comment (Paula-Kli) [reply backend/capellacollab/core/authentication/basic_auth.py:33]: alright good 👍

File: backend/capellacollab/sessions/hooks/t4c.py (hunk 4 above)
Comment (Paula-Kli) [reply backend/capellacollab/sessions/hooks/t4c.py:84]: good point

File: backend/capellacollab/users/tokens/crud.py (hunk 5 above)
Comment (Paula-Kli) [reply backend/capellacollab/users/tokens/crud.py:66]: I was thinking about it whether a programming paradigm exists but I don't know. I have often times seen functions that delete objects return the deleted object. That meaning that deletion was successful. Do you have a strong opinion on that?

Comment (dominik003) [reply backend/capellacollab/users/tokens/crud.py:66]: For me we have two cases here. Either the deletion fails and then the `db.delete(...)` should raise an exception (i.e., we would never come to the return anyways) or everything works. So I don't really see a use case for returning the deleted object. How do the places where you've seen this handle this?

File: frontend/src/app/general/auth/basic_auth_token/basic_auth_token.component.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply frontend/src/app/general/auth/basic_auth_token/basic_auth_token.component.html:34]: now it looks like
Bildschirmfoto 2023-09-22 um 11 10 08

was that what you imagined it to look like? I am a bit wondering about the mat-card because now it is just another card as all the tokens shown as well but maybe it is also fine

**Comment by Paula-Kli** [comment]
For me there are still two open issues in this PR:

1. in the tests I wrote a comment because due to monkey patching one test always succeeds but does not actually test correctly
2. The question @dominik003 raised regarding oauth and azure authentication and how to handle the basic auth there

File: backend/capellacollab/users/tokens/crud.py (hunk 5 above)
Comment (Paula-Kli) [reply backend/capellacollab/users/tokens/crud.py]: I deleted it

File: frontend/src/app/general/auth/basic_auth_token/basic_auth_token.component.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply frontend/src/app/general/auth/basic_auth_token/basic_auth_token.component.html]: changed it to
Bildschirmfoto 2023-09-25 um 09 03 45

File: docs/user/docs/tokens.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+all tokens that belong to your user. Clicking on the trash symbol you can
+delete a token which will no longer be valid to authenticate.
+
+## Token Usage
+
+The token created is a basic authentication token. There are different ways to
+authenticate with that against the Collaboration Manager API. One example is:
+
+```zsh
+curl --basic -u yourUsername:yourPassword https://baseurl/api/v1/users/current/tokens
```
Comment (MoritzWeber0) [inline docs/user/docs/tokens.md:42]:
```suggestion
curl --basic -u <username>:<token> https://<baseURL>/api/v1/users/current/tokens
```

Comment (MoritzWeber0) [inline docs/user/docs/tokens.md:42]: This just reminded me that there is no public API documentation yet, added a new issue for it: https://github.com/DSD-DBS/capella-collab-manager/issues/1066. 

But this is out of scope for this PR.

File: docs/user/docs/tokens.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+or to work with the diagram cache
+
+```python
+capellambse.model.MelodyModel(
+          path="path to the model on your machine",
+          diagram_cache={
+            "path": "https://baseurl/api/projects/[yourProjectSlug]/[yourModelName]/diagrams/%s",
+            "username": "yourUsername",
+            "password": "yourPassword",
```
Comment (MoritzWeber0) [inline docs/user/docs/tokens.md:53]:
```suggestion
            "password": "<token>",
```

File: docs/user/docs/tokens.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+```
+
+or to work with the diagram cache
+
+```python
+capellambse.model.MelodyModel(
+          path="path to the model on your machine",
+          diagram_cache={
+            "path": "https://baseurl/api/projects/[yourProjectSlug]/[yourModelName]/diagrams/%s",
+            "username": "yourUsername",
```
Comment (MoritzWeber0) [inline docs/user/docs/tokens.md:52]:
```suggestion
            "username": "<username>",
```

File: docs/user/docs/tokens.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+curl --basic -u yourUsername:yourPassword https://baseurl/api/v1/users/current/tokens
+```
+
+or to work with the diagram cache
+
+```python
+capellambse.model.MelodyModel(
+          path="path to the model on your machine",
+          diagram_cache={
+            "path": "https://baseurl/api/projects/[yourProjectSlug]/[yourModelName]/diagrams/%s",
```
Comment (MoritzWeber0) [inline docs/user/docs/tokens.md:51]:
```suggestion
            "path": "https://<baseURL>/api/projects/[yourProjectSlug]/[yourModelSlug]/diagrams/%s",
```

File: docs/user/docs/tokens.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+## Token Usage
+
+The token created is a basic authentication token. There are different ways to
+authenticate with that against the Collaboration Manager API. One example is:
+
+```zsh
+curl --basic -u yourUsername:yourPassword https://baseurl/api/v1/users/current/tokens
+```
+
+or to work with the diagram cache
```
Comment (MoritzWeber0) [inline docs/user/docs/tokens.md:45]: Please include a bit more information. What is capellambse? The context is missing. 
Also add a link to the capellambse Github repository.

File: docs/user/docs/tokens.md [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+## Token Creation
+
+To create a Token you can go to Profile > Token and insert a short description
+to create a token.
+
+ <!-- prettier-ignore -->
+
+!!! info The password which is generated can only be copied. Make sure you save
+it - you won't be able to access it again
```
Comment (MoritzWeber0) [inline docs/user/docs/tokens.md:20]: This is broken: 
image

File: docs/user/docs/tokens.md [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+ ~ SPDX-License-Identifier: Apache-2.0
+ -->
+
+# Token Authentication
+
+To authenticate against the API you can either use a Bearer token (which the
+browser usually uses) or authenticate with a longer lived basic authentication
+token which can be used e.g. in scripts.
+
+## Token Creation
```
Comment (MoritzWeber0) [inline docs/user/docs/tokens.md:12]:
```suggestion
## Personal access token creation
```

File: docs/user/docs/tokens.md [hunk 15]
```diff
@@ -0,0 +1,55 @@
+<!--
+ ~ SPDX-FileCopyrightText: Copyright DB Netz AG and the capella-collab-manager contributors
+ ~ SPDX-License-Identifier: Apache-2.0
+ -->
+
+# Token Authentication
+
+To authenticate against the API you can either use a Bearer token (which the
+browser usually uses) or authenticate with a longer lived basic authentication
+token which can be used e.g. in scripts.
```
Comment (MoritzWeber0) [inline docs/user/docs/tokens.md:10]:
```suggestion
token (PAT) which can be used e.g. in scripts.
```

File: docs/user/docs/tokens.md (hunk 14 above)
Comment (MoritzWeber0) [inline docs/user/docs/tokens.md:12]:
```suggestion
## PAT Creation
```
Let's call it PAT. Otherwise it's not clear whether we are talking about basic authentication tokens or Bearer tokens.

File: docs/user/docs/tokens.md [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+to create a token.
+
+ <!-- prettier-ignore -->
+
+!!! info The password which is generated can only be copied. Make sure you save
+it - you won't be able to access it again
+
+## Token Scope
+
+Basic authentication token have the same scope as the bearer token.
```
Comment (MoritzWeber0) [inline docs/user/docs/tokens.md:24]:
```suggestion
Basic authentication token have the same scope as the user who created it. It is therefore important that you never pass on the token and treat it responsibly. If you feel that the token has been lost, revoke it immediately and inform your Systems Engineering Toolchain team.
```

The typical user has no idea what a Bearer token is and wouldn't understand it.

File: docs/user/docs/tokens.md [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+
+<!-- prettier-ignore -->
+!!! warning
+    As currently Basic Authentication token have the same scope as you logged in your browser they can act as your user. Please do not share the password and if the information gets lost please revoke the token as soon as possible.
```
Comment (MoritzWeber0) [inline docs/user/docs/tokens.md:28]: Not needed anymore if you add the content of the last comment.

File: docs/user/docs/tokens.md [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+or to work with the diagram cache
+
+```python
+capellambse.model.MelodyModel(
+          path="path to the model on your machine",
+          diagram_cache={
+            "path": "https://baseurl/api/projects/[yourProjectSlug]/[yourModelName]/diagrams/%s",
+            "username": "yourUsername",
+            "password": "yourPassword",
+          })
```
Comment (MoritzWeber0) [inline docs/user/docs/tokens.md:54]:
```suggestion
capellambse.model.MelodyModel(
  path="path to the model on your machine",
  diagram_cache={
    "path": "https://baseurl/api/projects/[yourProjectSlug]/[yourModelName]/diagrams/%s",
    "username": "yourUsername",
    "password": "yourPassword",
  }
)
```

Wrong indentation and the closing bracket should be in a new line.

**Review by MoritzWeber0 — CHANGES_REQUESTED** [review]
This is it so far for the frontend + docs. 
I'll continue with the review of the backend after my lunch break.

File: backend/capellacollab/core/authentication/basic_auth.py [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+                        "reason": "The Signature of the token is expired. Please request a new access token.",
+                    },
+                    headers={"WWW-Authenticate": "Bearer, Basic"},
+                )
+                if self.auto_error:
+                    raise error
+                return None, error
+        return self.get_username(credentials), None
+
+    def get_username(self, credentials: security.HTTPBasicCredentials) -> str:
```
Comment (MoritzWeber0) [inline backend/capellacollab/core/authentication/basic_auth.py:69]: I didn't like the two different functions for `get_username` (as injectable and not as injectable). I've created a commit to merge those, please have look here: https://github.com/DSD-DBS/capella-collab-manager/pull/1005/commits/df6cc93f469158a5c48c6aaec2072fc7604155eb

File: backend/capellacollab/core/authentication/injectables.py [hunk 20]
```diff
@@ -3,19 +3,95 @@
...
+        return jwt_username
+
+    basic_username, basic_error = basic
+    if basic_username:
+        return basic_username
+
+    if jwt_error:
+        raise jwt_error
+    if basic_error:
+        raise basic_error
```
Comment (MoritzWeber0) [inline backend/capellacollab/core/authentication/injectables.py:88]:
```suggestion
```suggestion
    jwt_username, jwt_error = jwt
    basic_username, basic_error = basic
    
    if username := jwt_username or basic_username:
        return username
    if error := jwt_error or basic_error:
        raise error
```
```

File: backend/capellacollab/users/tokens/routes.py [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+        fastapi.Depends(
+            auth_injectables.RoleVerification(
+                required_role=users_models.Role.USER
+            )
+        )
+    ]
+)
+
+
+@router.post("/current/token")
```
Comment (MoritzWeber0) [inline backend/capellacollab/users/tokens/routes.py:25]:
```suggestion
@router.post("/current/tokens")
```

RESTful APIs usually use the plural form for resources: https://learn.microsoft.com/en-us/azure/architecture/best-practices/api-design

File: backend/capellacollab/users/tokens/routes.py [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+def create_token_for_user(
+    user: users_models.DatabaseUser = fastapi.Depends(
+        user_injectables.get_own_user
+    ),
+    db: orm.Session = fastapi.Depends(database.get_db),
+    description: str = fastapi.Body(),
```
Comment (MoritzWeber0) [inline backend/capellacollab/users/tokens/routes.py:31]: We should accept an object with attributes "description" and "expiration_date" here. This enabled us to add more attributes in the future without a breaking change.

File: backend/capellacollab/users/tokens/routes.py [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+):
```
Comment (MoritzWeber0) [inline backend/capellacollab/users/tokens/routes.py:32]: Missing type annotation.

File: backend/capellacollab/users/tokens/routes.py [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+    return password
+
+
+@router.get("/current/tokens")
+def get_all_token_of_user(
+    user: users_models.DatabaseUser = fastapi.Depends(
+        user_injectables.get_own_user
+    ),
+    db: orm.Session = fastapi.Depends(database.get_db),
+):
```
Comment (MoritzWeber0) [inline backend/capellacollab/users/tokens/routes.py:43]: Missing type annotation.

File: backend/capellacollab/users/tokens/routes.py [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+
+
+@router.delete("/current/token/{id}")
+def delete_token_for_user(
+    id: int,
+    user: users_models.DatabaseUser = fastapi.Depends(
+        user_injectables.get_own_user
+    ),
+    db: orm.Session = fastapi.Depends(database.get_db),
+):
```
Comment (MoritzWeber0) [inline backend/capellacollab/users/tokens/routes.py:54]: Missing type annotation.

File: backend/capellacollab/users/tokens/routes.py [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+    token_list = crud.get_token_by_user(db, user.id)
```
Comment (MoritzWeber0) [inline backend/capellacollab/users/tokens/routes.py:55]: This can be done with dependency injection: get_existing_user_token.

File: backend/capellacollab/users/tokens/routes.py [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+@router.post("/current/token")
+def create_token_for_user(
+    user: users_models.DatabaseUser = fastapi.Depends(
+        user_injectables.get_own_user
+    ),
+    db: orm.Session = fastapi.Depends(database.get_db),
+    description: str = fastapi.Body(),
+):
+    _, password = crud.create_token(db, user.id, description)
+    return password
```
Comment (MoritzWeber0) [inline backend/capellacollab/users/tokens/routes.py:34]: POST routes usually return the full resource, including the password. Not just one attribute of the resource.

**Review by MoritzWeber0 — CHANGES_REQUESTED** [review]
This is the second part of the review, I'm not yet finished with the backend review. The rest will follow on Wednesday.

**Review by MoritzWeber0 — CHANGES_REQUESTED** [review]
The general architecture looks good, a lot of it is just small comments. Thanks for the good feature, it will definitely help us to automate more in the future.

File: frontend/src/app/app.module.ts
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply frontend/src/app/app.module.ts:1]: oh ok - I do not have the issue but I have to rebase at some point anyhow

File: frontend/src/app/general/auth/basic-auth-token/basic-auth-token.component.html [hunk 28]
```diff
@@ -0,0 +1,78 @@
...
+<div class="input-container mb-10">
+  <div class="items-center">
+    <input
+      matInput
+      type="text"
+      [(ngModel)]="tokenDescription"
+      placeholder="Token description"
+      class="mr-10"
+    />
+    <button
```
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply frontend/src/app/general/auth/basic-auth-token/basic-auth-token.component.html:18]: I really like the idea of that being a hint close to the form field but at the same time they are so small - so I for now have the note twice and maybe you have an opinion about it

File: frontend/src/app/general/auth/basic-auth-token/basic-auth-token.component.ts [hunk 29]
```diff
@@ -0,0 +1,56 @@
...
+
+  ngOnInit() {
+    this.tokenService.loadTokens();
+  }
+
+  createNewToken(description?: string) {
+    if (!description) {
+      description = this.tokenDescription;
+    }
+    description = 'Token-overview-' + description;
```
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply frontend/src/app/general/auth/basic-auth-token/basic-auth-token.component.ts:34]: I thought of the other PR where you had only one button for creation of the token and I think for context of where it was created that might be helpful

File: frontend/src/app/general/auth/basic-auth-token/basic-auth-token.component.ts [hunk 30]
```diff
@@ -0,0 +1,56 @@
...
+    this.tokenService
+      .createToken(description)
+      .subscribe((token) => (this.password = token.replaceAll('"', '')));
```
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply frontend/src/app/general/auth/basic-auth-token/basic-auth-token.component.ts:37]: what we get back is a string looking like `"mystring"` and I wanted to remove the `"`

File: docs/user/mkdocs.yml [hunk 31]
```diff
@@ -9,6 +9,7 @@ nav:
   - Getting Started:
       - Overview: getting-started/overview.md
       - Introduction to Capella: getting-started/capella-intro.md
+  - Authentication: tokens.md
```
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply docs/user/mkdocs.yml:12]: I put it between User Management and `Release notes` as I think that the `Release notes` and `About` both are organizational more than content of the user docs - but then I am totally fine to put the tokens a bit more to the bottom

File: frontend/src/app/general/auth/basic-auth-token/basic-auth-token.component.html [hunk 32]
```diff
@@ -0,0 +1,78 @@
...
+      mat-raised-button
+    >
+      Create Token
+    </button>
+  </div>
+  <div *ngIf="password" class="mb-10">
+    <mat-card>
+      Generated Token:
+      <span
+        class="font-mono border cursor-copy mr-1"
```
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply frontend/src/app/general/auth/basic-auth-token/basic-auth-token.component.html:32]: but then it gets incompatible with the team4capella token UI?

File: frontend/src/app/general/auth/basic-auth-token/basic-auth-token.component.ts (hunk 29 above)
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply frontend/src/app/general/auth/basic-auth-token/basic-auth-token.component.ts:34]: That's a good thought and I like it - but we should do it explicitly then. Introduce a new field "source", which states where the token was created.

File: frontend/src/app/general/auth/basic-auth-token/basic-auth-token.component.ts (hunk 30 above)
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply frontend/src/app/general/auth/basic-auth-token/basic-auth-token.component.ts:37]: I suppose that you get JSON back then? Make sure that the response is parsed properly (a properly parsed response is from type string and has no quotation marks anymore). We don't want to use our own parsers when a parser is already available.

File: frontend/src/app/general/auth/basic-auth-token/basic-auth-token.component.html (hunk 32 above)
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply frontend/src/app/general/auth/basic-auth-token/basic-auth-token.component.html:32]: At some point, we need to ensure consistency, either within this PR or in a separate one. However, striving for consistency should not prevent us from making appropriate adjustments. If consistency is crucial, consider modifying the T4C tokens as well.

File: backend/capellacollab/users/tokens/routes.py (hunk 27 above)
Comment (Paula-Kli) [reply backend/capellacollab/users/tokens/routes.py:40]: how do I include the password in a good way as we do not safe that in the database

File: backend/capellacollab/sessions/hooks/jupyter.py [hunk 33]
```diff
@@ -63,7 +63,7 @@ def configuration_hook(
             "CSP_ORIGIN_HOST": f"{self._general_conf.get('scheme')}://{self._general_conf.get('host')}:{self._general_conf.get('port')}",
         }
 
-        volumes = self._get_project_share_volume_mounts(db, token, tool)
+        volumes = self._get_project_share_volume_mounts(db, username, tool)
```
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply backend/capellacollab/sessions/hooks/jupyter.py:66]: why would you choose to take `user.name` here? I think the token was injected before in order to get the username of the person requesting thus I changed it to the username injectable derived from the token now. Are they always the same and we can even remove the username at all?

[the remark this answers is not in the document]
Comment (Paula-Kli) [reply backend/capellacollab/sessions/hooks/jupyter.py:66]: alright you saw the same and the comment above answers that question already

File: backend/capellacollab/core/authentication/basic_auth.py [hunk 34]
```diff
@@ -0,0 +1,70 @@
...
+                raise error
+            return None, error
+        with database.SessionLocal() as session:
+            user = user_crud.get_user_by_name(session, credentials.username)
+            token_data = None
+            if user:
+                token_data = token_crud.get_token(
+                    session, credentials.password, user.id
+                )
+            if not token_data or not user or token_data.user_id != user.id:
```
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply backend/capellacollab/core/authentication/basic_auth.py:40]: I think it is in order to prevent me to authenticate as `paula` with your `token`. Thus we have to ensure that the given username provided matches the username that has this password

File: backend/capellacollab/core/authentication/basic_auth.py [hunk 35]
```diff
@@ -0,0 +1,70 @@
...
+                detail="Not authenticated",
+                headers={"WWW-Authenticate": "Bearer, Basic"},
+            )
+            if self.auto_error:
+                raise error
+            return None, error
+        with database.SessionLocal() as session:
+            user = user_crud.get_user_by_name(session, credentials.username)
+            token_data = None
+            if user:
```
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply backend/capellacollab/core/authentication/basic_auth.py:36]: but the later check groups the errors in order to have less code duplication doesn't it?

File: backend/capellacollab/users/tokens/crud.py [hunk 36]
```diff
@@ -0,0 +1,65 @@
...
+
+from capellacollab.core import credentials
+
+from . import models
+
+
+def create_token(
+    db: orm.Session, user_id: int, description: str
+) -> tuple[models.DatabaseUserTokenModel, str]:
+    password = credentials.generate_password(32)
```
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply backend/capellacollab/users/tokens/crud.py:23]: I think `collabmanager` could be more descriptive

File: backend/capellacollab/users/tokens/models.py [hunk 37]
```diff
@@ -0,0 +1,21 @@
...
+from capellacollab.core import database
+
+
+class DatabaseUserTokenModel(database.Base):
+    __tablename__ = "basic_auth_token"
+
+    id: orm.Mapped[int] = orm.mapped_column(
+        primary_key=True, index=True, autoincrement=True
+    )
+    user_id: orm.Mapped[int] = orm.mapped_column(sa.ForeignKey("users.id"))
```
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply backend/capellacollab/users/tokens/models.py:40]: I put this in the description of the table `sa.ForeignKeyConstraint(["user_id"], ["users.id"], ondelete="CASCADE"),` but it does not work - even though reading the [documentation](https://docs.sqlalchemy.org/en/20/core/constraints.html#sqlalchemy.schema.ForeignKeyConstraint.__init__) I thought it could - do you have an idea how I fix this?

[the remark this answers is not in the document]
Comment (Paula-Kli) [reply backend/capellacollab/users/tokens/models.py:43]: done thanks

File: backend/capellacollab/users/tokens/crud.py (hunk 36 above)
[the remark this answers is not in the document]
Comment (MoritzWeber0) [reply backend/capellacollab/users/tokens/crud.py]: I don't think that we need a descriptive name. And `collabmanager` is pretty long for a prefix. I'd stick to a short (three char) prefix, like Github has it.

[the remark this answers is not in the document]
Comment (Paula-Kli) [reply backend/capellacollab/users/tokens/crud.py]: I think you are referring to gitlab as my github token starts with `github_pat`

File: backend/tests/users/test_tokens.py [hunk 38]
```diff
@@ -0,0 +1,101 @@
...
+    "expiration_date": str(datetime.datetime.now()),
+    "description": "test_token",
+    "source": "test source",
+}
+
+
+@responses.activate
+@pytest.mark.usefixtures("user")
+def test_create_user_token(client: testclient.TestClient):
+    client.get(f"/api/v1/users/current")
```
[the remark this answers is not in the document]
Comment (Paula-Kli) [reply backend/tests/users/test_tokens.py:28]: good catch!

**Review by MoritzWeber0 — APPROVED** [review]
